

<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Usage &mdash; scATS 0.5.4 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
      <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=1ab9856b" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=c3dca1ed"></script>
      <script src="../_static/doctools.js?v=888ff710"></script>
      <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
      <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
      <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Disease model" href="Disease%20model.html" />
    <link rel="prev" title="Installation" href="Installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            scATS
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#tss-inference-and-quantification">TSS inference and quantification</a></li>
<li class="toctree-l2"><a class="reference internal" href="#finding-differentially-expressed-atss">Finding differentially expressed ATSs</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#finding-markers-of-all-group">Finding markers of all group</a></li>
<li class="toctree-l3"><a class="reference internal" href="#finding-markers-of-given-group">Finding markers of given group</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#calculating-psi">Calculating PSI</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sashimi-plots">Sashimi plots</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Disease%20model.html">Disease model</a></li>
<li class="toctree-l1"><a class="reference internal" href="Getting%20Help.html">Getting Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="Citing%20scATS.html">Citing scATS</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">scATS</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Usage</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/md/Usage.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section class="tex2jax_ignore mathjax_ignore" id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this heading"></a></h1>
<p><strong>scATS</strong> can analyze both single-end and paired-end 5’-end scRNA-seq data, enabling direct quantification using Seurat objects, while incorporating <strong>RNA degradation</strong> (RD) modeling through expectation-maximization (EM).</p>
<p>The main functions of scATS are: <code class="docutils literal notranslate"><span class="pre">TSSCDF</span></code>, <code class="docutils literal notranslate"><span class="pre">FindMarkers</span></code>, <code class="docutils literal notranslate"><span class="pre">psi</span></code>, <code class="docutils literal notranslate"><span class="pre">Sashimi</span></code></p>
<section id="tss-inference-and-quantification">
<h2>TSS inference and quantification<a class="headerlink" href="#tss-inference-and-quantification" title="Permalink to this heading"></a></h2>
<p>The input files include:</p>
<ul class="simple">
<li><p>Seurat object (R object)</p></li>
<li><p>Alignment file (bam file)</p></li>
<li><p>Annotation file (gtf file)</p></li>
</ul>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># load packages</span>
<span class="nf">suppressMessages</span><span class="p">({</span>
<span class="w">    </span><span class="nf">library</span><span class="p">(</span><span class="n">scATS</span><span class="p">)</span>
<span class="w">    </span><span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
<span class="p">})</span>

<span class="c1"># load input files</span>
<span class="n">seu</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">readRDS</span><span class="p">(</span><span class="s">&quot;demo_seurat.Rds&quot;</span><span class="p">)</span>
<span class="n">Genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rownames</span><span class="p">(</span><span class="n">seu</span><span class="p">)</span>
<span class="n">gtfFile</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;demo.gtf&quot;</span><span class="p">)</span>
<span class="n">bams</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;demo.bam&quot;</span>
<span class="nf">file.exists</span><span class="p">(</span><span class="n">bams</span><span class="p">)</span>

<span class="c1"># quantification using wrapper function</span>
<span class="n">scats</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">TSSCDF</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">seu</span><span class="p">,</span><span class="w"> </span><span class="n">bam</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bams</span><span class="p">,</span><span class="w"> </span><span class="n">gtf</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gtfFile</span><span class="p">,</span><span class="w"> </span><span class="n">genes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Genes</span><span class="p">,</span><span class="w"> </span><span class="n">verbose</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
<span class="n">scats</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>class: scATSDataSet
dim: 2043 2000
metadata(2): version parameters
assays(3): counts psi theta
rownames(2043): OR4F5@1:69063:+ OR4F5@1:69071:+ ... OR2G6@1:248508175:+
ZNF672@1:248838224:+
rowData names(12): TSS gene_id ... alpha theta
colnames(2000): TGGACGCTCCTTCAAT-1 CGAGCACGTCAGAATA-1 ...
ATCGAGTAGCGGCTTC-1 CTAGAGTAGTGCCATT-1
colData names(4): orig.ident nCount_RNA nFeature_RNA Cell
</pre></div>
</div>
<p>The quantification results are stored in  <code class="docutils literal notranslate"><span class="pre">scats&#64;rowRanges</span></code> and contain the following columns:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>column name</p></th>
<th class="head"><p>content</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>seqnames</p></td>
<td><p>The chromosomal name of the gene to which TSS belongs.</p></td>
</tr>
<tr class="row-odd"><td><p>ranges</p></td>
<td><p>The genomic locus of TSS, where the growth rate is highest.</p></td>
</tr>
<tr class="row-even"><td><p>strand</p></td>
<td><p>The genomic strand of the gene to which TSS belongs.</p></td>
</tr>
<tr class="row-odd"><td><p>gene_id</p></td>
<td><p>The Ensembl ID of the gene to which TSS belongs.</p></td>
</tr>
<tr class="row-even"><td><p>gene_name</p></td>
<td><p>The HGNC Symbol of the gene to which TSS belongs.</p></td>
</tr>
<tr class="row-odd"><td><p>TSS</p></td>
<td><p>The ID of the TSS in the format of [seqnames:ranges:strand].</p></td>
</tr>
<tr class="row-even"><td><p>Region</p></td>
<td><p>The growth area of the sorted 5’-starts of read 1, or it can also be interpreted as the<br> TSS cluster. Another region must show a significant increase in the number of reads.</p></td>
</tr>
<tr class="row-odd"><td><p>PSI</p></td>
<td><p>The percent spliced in (PSI) of TSS.</p></td>
</tr>
<tr class="row-even"><td><p>Percent</p></td>
<td><p>The ratio of TSS reads.</p></td>
</tr>
<tr class="row-odd"><td><p>Annotated</p></td>
<td><p>The nearest annotated TSS locus that exists in the region, if it is NA, <br>indicates that there are no annotated TSS loci in the region.</p></td>
</tr>
<tr class="row-even"><td><p>Greedy</p></td>
<td><p>Greedy for the first TSS. If it is TRUE, it indicates that the first TSS is quantified<br> even if it does not meet the specified condition.”</p></td>
</tr>
<tr class="row-odd"><td><p>beta</p></td>
<td><p>The area under the cumulative distribution curve : close to 1, indicates no degradation,<br>while close to 0 indicates severe degradation.</p></td>
</tr>
<tr class="row-even"><td><p>alpha</p></td>
<td><p>The degradation index fitted using the EM algorithm: the larger the value, the more severe the degradation.</p></td>
</tr>
<tr class="row-odd"><td><p>theta</p></td>
<td><p>The PSI value fitted using the EM algorithm.</p></td>
</tr>
<tr class="row-even"><td><p>AllReads</p></td>
<td><p>The total number of reads used for quantification.</p></td>
</tr>
</tbody>
</table>
<p><em><strong>Note : All the following analyses are based on the <code class="docutils literal notranslate"><span class="pre">scats</span></code> object.</strong></em></p>
</section>
<section id="finding-differentially-expressed-atss">
<h2>Finding differentially expressed ATSs<a class="headerlink" href="#finding-differentially-expressed-atss" title="Permalink to this heading"></a></h2>
<section id="finding-markers-of-all-group">
<h3>Finding markers of all group<a class="headerlink" href="#finding-markers-of-all-group" title="Permalink to this heading"></a></h3>
<p><strong>Identifying ATS markers based on <code class="docutils literal notranslate"><span class="pre">θ</span></code> value.</strong></p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">DE</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">scATS</span><span class="o">::</span><span class="nf">FindMarkersByTheta</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">scats</span><span class="p">,</span><span class="w"> </span><span class="n">groupBy</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CellType&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">group1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">group2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">cores</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">)</span>
<span class="n">DE</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,]</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>      gene            TSS group1    theta1    theta2
    &lt;char&gt;         &lt;char&gt; &lt;char&gt;     &lt;num&gt;     &lt;num&gt;
1:  OR4F5 OR4F5@1:69063:+      A 0.2780345 0.2835192
2:  OR4F5 OR4F5@1:69063:+      B 0.2862817 0.2620500
    cell1    cell2   percent1   percent2          p
    &lt;int&gt;    &lt;int&gt;      &lt;num&gt;      &lt;num&gt;      &lt;num&gt;
1:      6       17   0.896861  1.2772352  0.9682337
2:     10       13   1.459854  0.9885932  0.5724791
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">DE</span></code> object contains following columns:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>column name</p></th>
<th class="head"><p>content</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>gene</p></td>
<td><p>The HGNC Symbol of the gene to which TSS belongs.</p></td>
</tr>
<tr class="row-odd"><td><p>TSS</p></td>
<td><p>The ID of the TSS in the format of [gene_name:seqnames:ranges:strand].</p></td>
</tr>
<tr class="row-even"><td><p>group1</p></td>
<td><p>The levels of group.</p></td>
</tr>
<tr class="row-odd"><td><p>theta1/2</p></td>
<td><p>The theta value of the TSS in the group.</p></td>
</tr>
<tr class="row-even"><td><p>cell1/2</p></td>
<td><p>The number of cell-type corresponding to group.</p></td>
</tr>
<tr class="row-odd"><td><p>percent1/2</p></td>
<td><p>The ratio of TSS reads.</p></td>
</tr>
<tr class="row-even"><td><p>p</p></td>
<td><p>The p-value from the Wilcoxon test for differences in psi values among groups.</p></td>
</tr>
</tbody>
</table>
<p><strong>Identifying ATS markers based on <code class="docutils literal notranslate"><span class="pre">ψ</span></code> value.</strong></p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">DE</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">scATS</span><span class="o">::</span><span class="nf">FindMarkers</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">scats</span><span class="p">,</span><span class="w"> </span><span class="n">groupBy</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CellType&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">group1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">group2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">cores</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="n">majorOnly</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">F</span><span class="p">)</span><span class="w"> </span>
<span class="n">DE</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,]</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>               TSS     G1     G2    n1    n2    N1    N2 Cells1 Cells2      PSI1
            &lt;char&gt; &lt;char&gt; &lt;char&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;  &lt;int&gt;  &lt;int&gt;     &lt;num&gt;
1: OR4F5@1:69071:+      B  Other    41    83   156   277    685   1315 0.2088708
2: OR4F5@1:69156:+      B  Other    40    90   156   277    685   1315 0.2169872
        PSI2 PseudobulkPSI1 PseudobulkPSI2  wald.test wilcox.test    prop.test
        &lt;num&gt;          &lt;num&gt;          &lt;num&gt;      &lt;num&gt;       &lt;num&gt;        &lt;num&gt;
1: 0.2488509      0.3492958      0.2088773 0.42343003  0.32714124 3.331726e-11
2: 0.2964099      0.1309859      0.2663185 0.06632703  0.08046456 7.544541e-12
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">DE</span></code> object contains following columns:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>column name</p></th>
<th class="head"><p>content</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>TSS</p></td>
<td><p>The ID of the TSS in the format of [gene_name:seqnames:ranges:strand].</p></td>
</tr>
<tr class="row-odd"><td><p>G1/2</p></td>
<td><p>The levels of group.</p></td>
</tr>
<tr class="row-even"><td><p>n1/2</p></td>
<td><p>The expression number of the TSS in the G1/2.</p></td>
</tr>
<tr class="row-odd"><td><p>N1/2</p></td>
<td><p>The expression number of the host gene in the G1/2.</p></td>
</tr>
<tr class="row-even"><td><p>Cells1/2</p></td>
<td><p>The number of cell-type corresponding to G1/2.</p></td>
</tr>
<tr class="row-odd"><td><p>PSI1/2</p></td>
<td><p>The average sum of all individual cell PSIs.</p></td>
</tr>
<tr class="row-even"><td><p>PseudobulkPSI1/2</p></td>
<td><p>The PSI value calculated by combining all reads and treating them as a pseudobulk sample.</p></td>
</tr>
<tr class="row-odd"><td><p>wald.test</p></td>
<td><p>The p-value from the Wald test for differences in psi values among groups.</p></td>
</tr>
<tr class="row-even"><td><p>wilcox.test</p></td>
<td><p>The p-value from the Wilcoxon test for differences in psi values among groups.</p></td>
</tr>
<tr class="row-odd"><td><p>prop.test</p></td>
<td><p>The p-value from the Proportion test for differences in psi values among groups.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="finding-markers-of-given-group">
<h3>Finding markers of given group<a class="headerlink" href="#finding-markers-of-given-group" title="Permalink to this heading"></a></h3>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">### based on θ value</span>
<span class="n">DE</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">scATS</span><span class="o">::</span><span class="nf">FindMarkersByTheta</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">scats</span><span class="p">,</span><span class="w"> </span><span class="n">groupBy</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CellType&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">group1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">group2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">cores</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">)</span>
<span class="c1">### based on  ψ value</span>
<span class="n">DE</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">scATS</span><span class="o">::</span><span class="nf">FindMarkers</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">scats</span><span class="p">,</span><span class="w"> </span><span class="n">groupBy</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CellType&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">group1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;A&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">group2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;B&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">cores</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="n">majorOnly</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">F</span><span class="p">,</span><span class="n">gene</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;OR4F5&quot;</span><span class="p">)</span><span class="w"> </span>
</pre></div>
</div>
<p>In addition, you can specify the host genes used in the calculation by setting the <code class="docutils literal notranslate"><span class="pre">gene</span></code> parameter.</p>
</section>
</section>
<section id="calculating-psi">
<h2>Calculating PSI<a class="headerlink" href="#calculating-psi" title="Permalink to this heading"></a></h2>
<p><strong>Calculating PSI based on <code class="docutils literal notranslate"><span class="pre">θ</span></code> value.</strong></p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">theta</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">scATS</span><span class="o">::</span><span class="nf">ThetaByGroup</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">scats</span><span class="p">,</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;OR4F5&quot;</span><span class="p">,</span><span class="n">groupBy</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CellType&quot;</span><span class="p">)</span>
<span class="n">theta</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,]</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Group       TSS     alpha     theta
&lt;char&gt;    &lt;char&gt;     &lt;num&gt;     &lt;num&gt;
1:      A 1:69156:+ 0.1061093 0.1718889
2:      A 1:69090:+ 0.3006347 0.2107304
</pre></div>
</div>
<p><strong>Calculating PSI based on <code class="docutils literal notranslate"><span class="pre">ψ</span></code> value.</strong></p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">psi</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">scATS</span><span class="o">::</span><span class="nf">psi</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">scats</span><span class="p">,</span><span class="w"> </span><span class="n">groupBy</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CellType&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">TSS</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;OR4F5@1:69071:+&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;OR4F5@1:69156:+&quot;</span><span class="p">))</span>
<span class="n">psi</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,]</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>                TSS groupBy Cells   N  n    mean        sd
1 OR4F5@1:69071:+       A   669 129 39 0.2562320 0.4188128
2 OR4F5@1:69071:+       B   685 156 41 0.2088708 0.3822363
          se         ci median Q1  Q3 mad iqr PseudobulkPSI
1 0.03687441 0.07296233      0  0 0.5   0 0.5     0.2067138
2 0.03060340 0.06045356      0  0 0.2   0 0.2     0.3492958
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">theta</span></code> or <code class="docutils literal notranslate"><span class="pre">psi</span></code> object contains following columns:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>column name</p></th>
<th class="head"><p>content</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Group/groupBy</p></td>
<td><p>The level of group.</p></td>
</tr>
<tr class="row-odd"><td><p>TSS</p></td>
<td><p>The ID of the TSS in the format of [gene_name:seqnames:ranges:strand].</p></td>
</tr>
<tr class="row-even"><td><p>alpha</p></td>
<td><p>The degradation index fitted using the EM algorithm.</p></td>
</tr>
<tr class="row-odd"><td><p>Cells</p></td>
<td><p>The number of cell-type corresponding to a given groups (The same applies to the following.).</p></td>
</tr>
<tr class="row-even"><td><p>N</p></td>
<td><p>The expression of the host gene.</p></td>
</tr>
<tr class="row-odd"><td><p>n</p></td>
<td><p>The expression of the TSS.</p></td>
</tr>
<tr class="row-even"><td><p>mean</p></td>
<td><p>The mean of PSI.</p></td>
</tr>
<tr class="row-odd"><td><p>sd</p></td>
<td><p>The standard deviation (std) of PSI.</p></td>
</tr>
<tr class="row-even"><td><p>se</p></td>
<td><p>The standard error (SE) of PSI.</p></td>
</tr>
<tr class="row-odd"><td><p>ci</p></td>
<td><p>The confidence interval (CI) of PSI.</p></td>
</tr>
<tr class="row-even"><td><p>median</p></td>
<td><p>The median of PSI.</p></td>
</tr>
<tr class="row-odd"><td><p>Q1</p></td>
<td><p>The first quartile (Q1) of PSI.</p></td>
</tr>
<tr class="row-even"><td><p>Q3</p></td>
<td><p>The third quartile (Q3) of PSI.</p></td>
</tr>
<tr class="row-odd"><td><p>mad</p></td>
<td><p>The median absolute deviation (MAD) of PSI.</p></td>
</tr>
<tr class="row-even"><td><p>iqr</p></td>
<td><p>The interquartile range (IQR) of PSI.</p></td>
</tr>
<tr class="row-odd"><td><p>theta/PseudobulkPSI</p></td>
<td><p>The PSI value calculated by combining all reads and treating them as a pseudobulk sample.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="sashimi-plots">
<h2>Sashimi plots<a class="headerlink" href="#sashimi-plots" title="Permalink to this heading"></a></h2>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Take SCNN1D gene as an example</span>
<span class="n">scATS</span><span class="o">::</span><span class="nf">Sashimi</span><span class="p">(</span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">scats</span><span class="p">,</span><span class="w"> </span>
<span class="w">               </span><span class="n">bam</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bams</span><span class="p">,</span>
<span class="w">               </span><span class="n">xlimit</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1280352</span><span class="p">,</span><span class="w"> </span><span class="m">1282325</span><span class="p">),</span>
<span class="w">               </span><span class="n">transcripts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;SCNN1D-201&quot;</span><span class="p">,</span><span class="s">&quot;SCNN1D-206&quot;</span><span class="p">,</span><span class="s">&quot;SCNN1D-205&quot;</span><span class="p">),</span>
<span class="w">               </span><span class="n">gtf</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gtfFile</span><span class="p">,</span><span class="w"> </span>
<span class="w">               </span><span class="n">gene</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;SCNN1D&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">               </span><span class="n">TSS</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1280452</span><span class="p">,</span><span class="w"> </span><span class="m">1281224</span><span class="p">,</span><span class="w"> </span><span class="m">1282125</span><span class="p">),</span><span class="w"> </span><span class="c1"># TSS位点</span>
<span class="w">               </span><span class="n">free_y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">T</span><span class="p">,</span><span class="c1">#是否scale</span>
<span class="w">               </span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="c1">#read部分字体大小</span>
<span class="w">               </span><span class="n">rel_height</span><span class="o">=</span><span class="m">0.9</span><span class="p">,</span><span class="w"> </span><span class="c1">#注释/read ，小于1 read部分比例更大</span>
<span class="w">               </span><span class="n">fill.color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;grey&quot;</span><span class="p">,</span>
<span class="w">               </span><span class="n">line.color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;red&quot;</span><span class="p">,</span>
<span class="w">               </span><span class="n">line.type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span><span class="w"> </span><span class="o">-&gt;</span><span class="w"> </span><span class="n">p</span>
<span class="n">p</span>
</pre></div>
</div>
<p><img alt="sashimi" src="../_images/image.png" /></p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">sessionInfo</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="w">    </span><span class="n">R</span><span class="w"> </span><span class="n">version</span><span class="w"> </span><span class="m">4.5</span><span class="n">.</span><span class="m">1</span><span class="w"> </span><span class="p">(</span><span class="m">2025-06-13</span><span class="p">)</span>
<span class="w">    </span><span class="n">Platform</span><span class="o">:</span><span class="w"> </span><span class="n">x86_64</span><span class="o">-</span><span class="n">conda</span><span class="o">-</span><span class="n">linux</span><span class="o">-</span><span class="n">gnu</span>
<span class="w">    </span><span class="n">Running</span><span class="w"> </span><span class="n">under</span><span class="o">:</span><span class="w"> </span><span class="n">CentOS</span><span class="w"> </span><span class="n">Linux</span><span class="w"> </span><span class="m">7</span><span class="w"> </span><span class="p">(</span><span class="n">Core</span><span class="p">)</span>

<span class="w">    </span><span class="n">Matrix</span><span class="w"> </span><span class="n">products</span><span class="o">:</span><span class="w"> </span><span class="n">default</span>
<span class="w">    </span><span class="n">BLAS</span><span class="o">/</span><span class="n">LAPACK</span><span class="o">:</span><span class="w"> </span><span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">data4</span><span class="o">/</span><span class="n">xuzijie</span><span class="o">/</span><span class="n">miniconda3</span><span class="o">/</span><span class="n">envs</span><span class="o">/</span><span class="n">r451</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">libopenblasp</span><span class="o">-</span><span class="n">r0.3.30.so</span><span class="p">;</span><span class="w">  </span><span class="n">LAPACK</span><span class="w"> </span><span class="n">version</span><span class="w"> </span><span class="m">3.12</span><span class="n">.</span><span class="m">0</span>

<span class="w">    </span><span class="n">locale</span><span class="o">:</span>
<span class="w">    </span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="n">LC_CTYPE</span><span class="o">=</span><span class="n">zh_CN.UTF</span><span class="m">-8</span><span class="w">       </span><span class="n">LC_NUMERIC</span><span class="o">=</span><span class="n">C</span><span class="w">               </span><span class="n">LC_TIME</span><span class="o">=</span><span class="n">zh_CN.UTF</span><span class="m">-8</span>
<span class="w">    </span><span class="p">[</span><span class="m">4</span><span class="p">]</span><span class="w"> </span><span class="n">LC_COLLATE</span><span class="o">=</span><span class="n">zh_CN.UTF</span><span class="m">-8</span><span class="w">     </span><span class="n">LC_MONETARY</span><span class="o">=</span><span class="n">zh_CN.UTF</span><span class="m">-8</span><span class="w">    </span><span class="n">LC_MESSAGES</span><span class="o">=</span><span class="n">zh_CN.UTF</span><span class="m">-8</span>
<span class="w">    </span><span class="p">[</span><span class="m">7</span><span class="p">]</span><span class="w"> </span><span class="n">LC_PAPER</span><span class="o">=</span><span class="n">zh_CN.UTF</span><span class="m">-8</span><span class="w">       </span><span class="n">LC_NAME</span><span class="o">=</span><span class="n">C</span><span class="w">                  </span><span class="n">LC_ADDRESS</span><span class="o">=</span><span class="n">C</span>
<span class="w">    </span><span class="p">[</span><span class="m">10</span><span class="p">]</span><span class="w"> </span><span class="n">LC_TELEPHONE</span><span class="o">=</span><span class="n">C</span><span class="w">             </span><span class="n">LC_MEASUREMENT</span><span class="o">=</span><span class="n">zh_CN.UTF</span><span class="m">-8</span><span class="w"> </span><span class="n">LC_IDENTIFICATION</span><span class="o">=</span><span class="n">C</span>

<span class="w">    </span><span class="n">time</span><span class="w"> </span><span class="n">zone</span><span class="o">:</span><span class="w"> </span><span class="n">America</span><span class="o">/</span><span class="n">New_York</span>
<span class="w">    </span><span class="n">tzcode</span><span class="w"> </span><span class="n">source</span><span class="o">:</span><span class="w"> </span><span class="nf">system </span><span class="p">(</span><span class="n">glibc</span><span class="p">)</span>

<span class="w">    </span><span class="n">attached</span><span class="w"> </span><span class="n">base</span><span class="w"> </span><span class="n">packages</span><span class="o">:</span>
<span class="w">    </span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="n">stats</span><span class="w">     </span><span class="n">graphics</span><span class="w">  </span><span class="n">grDevices</span><span class="w"> </span><span class="n">utils</span><span class="w">     </span><span class="n">datasets</span><span class="w">  </span><span class="n">methods</span><span class="w">   </span><span class="n">base</span>

<span class="w">    </span><span class="n">other</span><span class="w"> </span><span class="n">attached</span><span class="w"> </span><span class="n">packages</span><span class="o">:</span>
<span class="w">    </span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="n">scATS_0.5.5</span>

<span class="w">    </span><span class="n">loaded</span><span class="w"> </span><span class="n">via</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="nf">namespace </span><span class="p">(</span><span class="n">and</span><span class="w"> </span><span class="n">not</span><span class="w"> </span><span class="n">attached</span><span class="p">)</span><span class="o">:</span>
<span class="w">    </span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="n">RcppAnnoy_0.0.19</span><span class="w">            </span><span class="n">splines_4.5.1</span><span class="w">               </span><span class="n">later_1.3.2</span>
<span class="w">    </span><span class="p">[</span><span class="m">4</span><span class="p">]</span><span class="w"> </span><span class="n">bitops_1.0</span><span class="m">-7</span><span class="w">                </span><span class="n">tibble_3.2.1</span><span class="w">                </span><span class="n">R.oo_1.24.0</span>
<span class="w">    </span><span class="p">[</span><span class="m">7</span><span class="p">]</span><span class="w"> </span><span class="n">polyclip_1.10</span><span class="m">-0</span><span class="w">             </span><span class="n">graph_1.68.0</span><span class="w">                </span><span class="n">xts_0.13.1</span>
<span class="w">    </span><span class="p">[</span><span class="m">10</span><span class="p">]</span><span class="w"> </span><span class="n">XML_3.99</span><span class="m">-0.14</span><span class="w">               </span><span class="n">rpart_4.1</span><span class="m">-15</span><span class="w">                </span><span class="n">lifecycle_1.0.4</span>
<span class="w">    </span><span class="p">[</span><span class="m">13</span><span class="p">]</span><span class="w"> </span><span class="n">OrganismDbi_1.32.0</span><span class="w">          </span><span class="n">ensembldb_2.14.1</span><span class="w">            </span><span class="n">globals_0.16.3</span>
<span class="w">    </span><span class="p">[</span><span class="m">16</span><span class="p">]</span><span class="w"> </span><span class="n">lattice_0.22</span><span class="m">-9</span><span class="w">              </span><span class="n">MASS_7.3</span><span class="m">-58.3</span><span class="w">               </span><span class="n">backports_1.4.1</span>
<span class="w">    </span><span class="p">[</span><span class="m">19</span><span class="p">]</span><span class="w"> </span><span class="n">magrittr_2.0.3</span><span class="w">              </span><span class="n">Hmisc_4.6</span><span class="m">-0</span><span class="w">                 </span><span class="n">plotly_4.10.4</span>
<span class="w">    </span><span class="p">[</span><span class="m">22</span><span class="p">]</span><span class="w"> </span><span class="n">httpuv_1.6.15</span><span class="w">               </span><span class="n">Seurat_4.0.5</span><span class="w">                </span><span class="n">sctransform_0.3.2</span>
<span class="w">    </span><span class="p">[</span><span class="m">25</span><span class="p">]</span><span class="w"> </span><span class="n">spatstat.core_2.3</span><span class="m">-2</span><span class="w">         </span><span class="n">askpass_1.2.0</span><span class="w">               </span><span class="n">spatstat.sparse_3.0</span><span class="m">-3</span>
<span class="w">    </span><span class="p">[</span><span class="m">28</span><span class="p">]</span><span class="w"> </span><span class="n">reticulate_1.24</span><span class="w">             </span><span class="n">ggbio_1.38.0</span><span class="w">                </span><span class="n">cowplot_1.1.3</span>
<span class="w">    </span><span class="p">[</span><span class="m">31</span><span class="p">]</span><span class="w"> </span><span class="n">pbapply_1.7</span><span class="m">-2</span><span class="w">               </span><span class="n">DBI_1.2.2</span><span class="w">                   </span><span class="n">RColorBrewer_1.1</span><span class="m">-3</span>
<span class="w">    </span><span class="p">[</span><span class="m">34</span><span class="p">]</span><span class="w"> </span><span class="n">abind_1.4</span><span class="m">-5</span><span class="w">                 </span><span class="n">zlibbioc_1.36.0</span><span class="w">             </span><span class="n">Rtsne_0.17</span>
<span class="w">    </span><span class="p">[</span><span class="m">37</span><span class="p">]</span><span class="w"> </span><span class="n">GenomicRanges_1.42.0</span><span class="w">        </span><span class="n">mixtools_1.2.0</span><span class="w">              </span><span class="n">purrr_1.0.2</span>
<span class="w">    </span><span class="p">[</span><span class="m">40</span><span class="p">]</span><span class="w"> </span><span class="n">R.utils_2.11.0</span><span class="w">              </span><span class="n">AnnotationFilter_1.14.0</span><span class="w">     </span><span class="n">biovizBase_1.38.0</span>
<span class="w">    </span><span class="p">[</span><span class="m">43</span><span class="p">]</span><span class="w"> </span><span class="n">BiocGenerics_0.40.0</span><span class="w">         </span><span class="n">RCurl_1.98</span><span class="m">-1.5</span><span class="w">              </span><span class="n">nnet_7.3</span><span class="m">-16</span>
<span class="w">    </span><span class="p">[</span><span class="m">46</span><span class="p">]</span><span class="w"> </span><span class="n">VariantAnnotation_1.36.0</span><span class="w">    </span><span class="n">rappdirs_0.3.3</span><span class="w">              </span><span class="n">GenomeInfoDbData_1.2.4</span>
<span class="w">    </span><span class="p">[</span><span class="m">49</span><span class="p">]</span><span class="w"> </span><span class="n">IRanges_2.24.1</span><span class="w">              </span><span class="n">S4Vectors_0.28.1</span><span class="w">            </span><span class="n">ggrepel_0.9.5</span>
<span class="w">    </span><span class="p">[</span><span class="m">52</span><span class="p">]</span><span class="w"> </span><span class="n">irlba_2.3.5.1</span><span class="w">               </span><span class="n">listenv_0.9.1</span><span class="w">               </span><span class="n">spatstat.utils_3.1</span><span class="m">-3</span>
<span class="w">    </span><span class="p">[</span><span class="m">55</span><span class="p">]</span><span class="w"> </span><span class="n">goftest_1.2</span><span class="m">-3</span><span class="w">               </span><span class="n">fitdistrplus_1.1</span><span class="m">-6</span><span class="w">          </span><span class="n">parallelly_1.37.1</span>
<span class="w">    </span><span class="p">[</span><span class="m">58</span><span class="p">]</span><span class="w"> </span><span class="n">smoother_1.1</span><span class="w">                </span><span class="n">leiden_0.3.9</span><span class="w">                </span><span class="n">codetools_0.2</span><span class="m">-19</span>
<span class="w">    </span><span class="p">[</span><span class="m">61</span><span class="p">]</span><span class="w"> </span><span class="n">DelayedArray_0.16.3</span><span class="w">         </span><span class="n">xml2_1.3.6</span><span class="w">                  </span><span class="n">tidyselect_1.2.1</span>
<span class="w">    </span><span class="p">[</span><span class="m">64</span><span class="p">]</span><span class="w"> </span><span class="n">matrixStats_1.2.0</span><span class="w">           </span><span class="n">stats4_4.5.1</span><span class="w">                </span><span class="n">BiocFileCache_1.14.0</span>
<span class="w">    </span><span class="p">[</span><span class="m">67</span><span class="p">]</span><span class="w"> </span><span class="n">base64enc_0.1</span><span class="m">-3</span><span class="w">             </span><span class="n">GenomicAlignments_1.26.0</span><span class="w">    </span><span class="n">jsonlite_1.8.8</span>
<span class="w">    </span><span class="p">[</span><span class="m">70</span><span class="p">]</span><span class="w"> </span><span class="n">Formula_1.2</span><span class="m">-4</span><span class="w">               </span><span class="n">ggridges_0.5.3</span><span class="w">              </span><span class="n">survival_3.5</span><span class="m">-5</span>
<span class="w">    </span><span class="p">[</span><span class="m">73</span><span class="p">]</span><span class="w"> </span><span class="n">segmented_1.3</span><span class="m">-4</span><span class="w">             </span><span class="n">tools_4.5.1</span><span class="w">                 </span><span class="n">progress_1.2.3</span>
<span class="w">    </span><span class="p">[</span><span class="m">76</span><span class="p">]</span><span class="w"> </span><span class="n">ica_1.0</span><span class="m">-2</span><span class="w">                   </span><span class="n">Rcpp_1.0.12</span><span class="w">                 </span><span class="n">glue_1.7.0</span>
<span class="w">    </span><span class="p">[</span><span class="m">79</span><span class="p">]</span><span class="w"> </span><span class="n">gridExtra_2.3</span><span class="w">               </span><span class="n">xfun_0.43</span><span class="w">                   </span><span class="n">mgcv_1.8</span><span class="m">-42</span>
<span class="w">    </span><span class="p">[</span><span class="m">82</span><span class="p">]</span><span class="w"> </span><span class="n">TTR_0.24.3</span><span class="w">                  </span><span class="n">MatrixGenerics_1.2.1</span><span class="w">        </span><span class="n">GenomeInfoDb_1.30.1</span>
<span class="w">    </span><span class="p">[</span><span class="m">85</span><span class="p">]</span><span class="w"> </span><span class="n">dplyr_1.1.4</span><span class="w">                 </span><span class="n">BiocManager_1.30.22</span><span class="w">         </span><span class="n">fastmap_1.1.1</span>
<span class="w">    </span><span class="p">[</span><span class="m">88</span><span class="p">]</span><span class="w"> </span><span class="n">GGally_2.1.2</span><span class="w">                </span><span class="n">latticeExtra_0.6</span><span class="m">-29</span><span class="w">         </span><span class="n">fansi_1.0.6</span>
<span class="w">    </span><span class="p">[</span><span class="m">91</span><span class="p">]</span><span class="w"> </span><span class="n">openssl_2.1.1</span><span class="w">               </span><span class="n">digest_0.6.35</span><span class="w">               </span><span class="n">R6_2.5.1</span>
<span class="w">    </span><span class="p">[</span><span class="m">94</span><span class="p">]</span><span class="w"> </span><span class="n">mime_0.12</span><span class="w">                   </span><span class="n">colorspace_2.1</span><span class="m">-0</span><span class="w">            </span><span class="n">scattermore_1.2</span>
<span class="w">    </span><span class="p">[</span><span class="m">97</span><span class="p">]</span><span class="w"> </span><span class="n">tensor_1.5</span><span class="w">                  </span><span class="n">dichromat_2.0</span><span class="m">-0</span><span class="w">             </span><span class="n">jpeg_0.1</span><span class="m">-9</span>
<span class="w">    </span><span class="p">[</span><span class="m">100</span><span class="p">]</span><span class="w"> </span><span class="n">spatstat.data_3.0</span><span class="m">-4</span><span class="w">         </span><span class="n">biomaRt_2.46.3</span><span class="w">              </span><span class="n">RSQLite_2.2.9</span>
<span class="w">    </span><span class="p">[</span><span class="m">103</span><span class="p">]</span><span class="w"> </span><span class="n">R.methodsS3_1.8.1</span><span class="w">           </span><span class="n">utf8_1.2.4</span><span class="w">                  </span><span class="n">tidyr_1.3.1</span>
<span class="w">    </span><span class="p">[</span><span class="m">106</span><span class="p">]</span><span class="w"> </span><span class="n">generics_0.1.3</span><span class="w">              </span><span class="n">data.table_1.15.4</span><span class="w">           </span><span class="n">rtracklayer_1.50.0</span>
<span class="w">    </span><span class="p">[</span><span class="m">109</span><span class="p">]</span><span class="w"> </span><span class="n">prettyunits_1.2.0</span><span class="w">           </span><span class="n">httr_1.4.7</span><span class="w">                  </span><span class="n">htmlwidgets_1.6.4</span>
<span class="w">    </span><span class="p">[</span><span class="m">112</span><span class="p">]</span><span class="w"> </span><span class="n">uwot_0.1.16</span><span class="w">                 </span><span class="n">pkgconfig_2.0.3</span><span class="w">             </span><span class="n">gtable_0.3.5</span>
<span class="w">    </span><span class="p">[</span><span class="m">115</span><span class="p">]</span><span class="w"> </span><span class="n">blob_1.2.4</span><span class="w">                  </span><span class="n">lmtest_0.9</span><span class="m">-39</span><span class="w">               </span><span class="n">XVector_0.30.0</span>
<span class="w">    </span><span class="p">[</span><span class="m">118</span><span class="p">]</span><span class="w"> </span><span class="n">htmltools_0.5.8.1</span><span class="w">           </span><span class="n">RBGL_1.66.0</span><span class="w">                 </span><span class="n">ProtGenerics_1.22.0</span>
<span class="w">    </span><span class="p">[</span><span class="m">121</span><span class="p">]</span><span class="w"> </span><span class="n">SeuratObject_4.0.4</span><span class="w">          </span><span class="n">scales_1.3.0</span><span class="w">                </span><span class="n">Biobase_2.50.0</span>
<span class="w">    </span><span class="p">[</span><span class="m">124</span><span class="p">]</span><span class="w"> </span><span class="n">png_0.1</span><span class="m">-8</span><span class="w">                   </span><span class="n">rstudioapi_0.16.0</span><span class="w">           </span><span class="n">knitr_1.46</span>
<span class="w">    </span><span class="p">[</span><span class="m">127</span><span class="p">]</span><span class="w"> </span><span class="n">reshape2_1.4.4</span><span class="w">              </span><span class="n">checkmate_2.0.0</span><span class="w">             </span><span class="n">nlme_3.1</span><span class="m">-162</span>
<span class="w">    </span><span class="p">[</span><span class="m">130</span><span class="p">]</span><span class="w"> </span><span class="n">curl_5.2.1</span><span class="w">                  </span><span class="n">cachem_1.0.8</span><span class="w">                </span><span class="n">zoo_1.8</span><span class="m">-9</span>
<span class="w">    </span><span class="p">[</span><span class="m">133</span><span class="p">]</span><span class="w"> </span><span class="n">stringr_1.5.1</span><span class="w">               </span><span class="n">KernSmooth_2.23</span><span class="m">-20</span><span class="w">          </span><span class="n">parallel_4.5.1</span>
<span class="w">    </span><span class="p">[</span><span class="m">136</span><span class="p">]</span><span class="w"> </span><span class="n">miniUI_0.1.1.1</span><span class="w">              </span><span class="n">foreign_0.8</span><span class="m">-81</span><span class="w">              </span><span class="n">AnnotationDbi_1.52.0</span>
<span class="w">    </span><span class="p">[</span><span class="m">139</span><span class="p">]</span><span class="w"> </span><span class="n">pillar_1.9.0</span><span class="w">                </span><span class="n">grid_4.5.1</span><span class="w">                  </span><span class="n">reshape_0.8.8</span>
<span class="w">    </span><span class="p">[</span><span class="m">142</span><span class="p">]</span><span class="w"> </span><span class="n">vctrs_0.6.5</span><span class="w">                 </span><span class="n">RANN_2.6.1</span><span class="w">                  </span><span class="n">VGAM_1.1</span><span class="m">-5</span>
<span class="w">    </span><span class="p">[</span><span class="m">145</span><span class="p">]</span><span class="w"> </span><span class="n">promises_1.3.0</span><span class="w">              </span><span class="n">dbplyr_2.5.0</span><span class="w">                </span><span class="n">xtable_1.8</span><span class="m">-4</span>
<span class="w">    </span><span class="p">[</span><span class="m">148</span><span class="p">]</span><span class="w"> </span><span class="n">cluster_2.1.4</span><span class="w">               </span><span class="n">htmlTable_2.3.0</span><span class="w">             </span><span class="n">GenomicFeatures_1.42.3</span>
<span class="w">    </span><span class="p">[</span><span class="m">151</span><span class="p">]</span><span class="w"> </span><span class="n">cli_3.6.3</span><span class="w">                   </span><span class="n">compiler_4.5.1</span><span class="w">              </span><span class="n">Rsamtools_2.6.0</span>
<span class="w">    </span><span class="p">[</span><span class="m">154</span><span class="p">]</span><span class="w"> </span><span class="n">rlang_1.1.4</span><span class="w">                 </span><span class="n">crayon_1.5.2</span><span class="w">                </span><span class="n">future.apply_1.11.2</span>
<span class="w">    </span><span class="p">[</span><span class="m">157</span><span class="p">]</span><span class="w"> </span><span class="n">mclust_6.0.0</span><span class="w">                </span><span class="n">plyr_1.8.9</span><span class="w">                  </span><span class="n">stringi_1.8.3</span>
<span class="w">    </span><span class="p">[</span><span class="m">160</span><span class="p">]</span><span class="w"> </span><span class="n">viridisLite_0.4.2</span><span class="w">           </span><span class="n">deldir_1.0</span><span class="m">-6</span><span class="w">                </span><span class="n">BiocParallel_1.24.1</span>
<span class="w">    </span><span class="p">[</span><span class="m">163</span><span class="p">]</span><span class="w"> </span><span class="n">munsell_0.5.1</span><span class="w">               </span><span class="n">Biostrings_2.58.0</span><span class="w">           </span><span class="n">lazyeval_0.2.2</span>
<span class="w">    </span><span class="p">[</span><span class="m">166</span><span class="p">]</span><span class="w"> </span><span class="n">spatstat.geom_3.2</span><span class="m">-9</span><span class="w">         </span><span class="n">Matrix_1.5</span><span class="m">-3</span><span class="w">                </span><span class="n">BSgenome_1.58.0</span>
<span class="w">    </span><span class="p">[</span><span class="m">169</span><span class="p">]</span><span class="w"> </span><span class="n">hms_1.1.3</span><span class="w">                   </span><span class="n">patchwork_1.2.0</span><span class="w">             </span><span class="n">bit64_4.0.5</span>
<span class="w">    </span><span class="p">[</span><span class="m">172</span><span class="p">]</span><span class="w"> </span><span class="n">future_1.33.2</span><span class="w">               </span><span class="n">ggplot2_3.5.1</span><span class="w">               </span><span class="n">shiny_1.8.1.1</span>
<span class="w">    </span><span class="p">[</span><span class="m">175</span><span class="p">]</span><span class="w"> </span><span class="n">SummarizedExperiment_1.20.0</span><span class="w"> </span><span class="n">kernlab_0.9</span><span class="m">-29</span><span class="w">              </span><span class="n">ROCR_1.0</span><span class="m">-11</span>
<span class="w">    </span><span class="p">[</span><span class="m">178</span><span class="p">]</span><span class="w"> </span><span class="n">igraph_1.5.1</span><span class="w">                </span><span class="n">memoise_2.0.1</span><span class="w">               </span><span class="n">bit_4.0.5</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Disease%20model.html" class="btn btn-neutral float-right" title="Disease model" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, LuChenLab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-XXXXXXXXXX"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'G-XXXXXXXXXX', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>